cpp/arcticdb/storage/mongo/mongo_client_wrapper.hpp

```diff
@@ -0,0 +1,78 @@
+/* Copyright 2023 Man Group Operations Limited
+ *
+ * Use of this software is governed by the Business Source License 1.1 included in the file licenses/BSL.txt.
+ *
+ * As of the Change Date specified in that file, in accordance with the Business Source License, use of this software will be governed by the Apache License, version 2.0.
+ */
+
+#pragma once
+#include <fmt/format.h>
+
+#include <arcticdb/storage/storage.hpp>
+#include <arcticdb/entity/protobufs.hpp>
+
+namespace arcticdb::storage::mongo {
+
+template<class Output, class Result>
+struct MongoResult{
+    std::variant<Output, Result> result;
+
+    [[nodiscard]] bool is_success() const {
+        return std::holds_alternative<Output>(result);
+    }
+
+    Result& get_error(){
+        return std::get<Result>(result);
+    }
+    Output& get_output(){
+        return std::get<Output>(result);
+    }
+};
+
+class MongoClientWrapper {
+public:
+    virtual ~MongoClientWrapper() = default;
+
+    virtual bool write_segment(
+            const std::string &database_name,
+            const std::string &collection_name,
+            storage::KeySegmentPair&& kv) = 0;
+
+    virtual std::optional<int> update_segment(
```

Comment:
Additionally I think I prefer returning something like a struct `UpdateResult{std::optional<int>modified_count;}` which makes it clear what the returned int is without looking at documentation. 

Potentially the `MongoResult` type created above could also be a nicer way to signal a failure rather just returning an `std::optional`

Still, I don't feel too strongly about this, so if you think this will clutter code too much it's ok to leave as documentation.

Reply:
Fixed in [6100dd8](https://github.com/man-group/ArcticDB/pull/1383/commits/6100dd824d3fe48bc32399c377f1b477195c7654)